linux/src/threadpool_linux.c

```diff
@@ -545,15 +639,20 @@ int threadpool_timer_start(THREADPOOL_HANDLE threadpool, uint32_t start_delay_ms
     }
     else
     {
-        TIMER_INSTANCE* timer_instance = malloc(sizeof(TIMER_INSTANCE));
+        TIMER_INSTANCE* timer_instance = get_next_timer_instance(threadpool);
         if (timer_instance == NULL)
         {
-            LogError("Failure allocating Timer Instance");
+            LogError("Failure getting Timer Instance all timers are being used");
+        }
+        else if (clock_gettime(CLOCK_REALTIME, &timer_instance->start_time) == -1)
+        {
+            LogError("Failure clock_gettime");
         }
         else
         {
             timer_instance->work_function = work_function;
             timer_instance->work_function_ctx = work_function_ctx;
```

Comment:
Ok so thinking about any more races here between a thread (T1) calling timer start, and another thread (T2) which is an extra timer callback from a deleted timer.

T1 == threadpool_timer_start
T1:1 set state ENABLED
T1:2 set start_time
T1:3 set timer_trigger 0
T1:4 timer_create

T2 == on_timer_callback
T2:1 check state ENABLED
T2:2 check timer_trigger 0
T2:3 set timer_trigger 1
T2:4 get time
T2:5 compare times
T2:6 call callback

We have a race if T1:1, T1:2 happens, then T2:1, T2:2 happens (which will not see timer_trigger as 0 so it will just call the callback from the else branch)

Otherwise I think the order is correct.

Why don't we get rid of `timer_trigger`. Instead we can have the states ENABLING, ENABLED, DISABLED. T1:1 sets to ENABLING first, then it can change to ENABLED before T1:4.
That way T2 will not see it ENABLED until we are done with all that setup.

This would solve the other comment I was going to make which is I don't know if setting `start_time` with `clock_gettime` is atomic. If we essentially "lock" changing `start_time` under a critical section (state==ENABLING) then we won't ever see partial updates from the other thread.